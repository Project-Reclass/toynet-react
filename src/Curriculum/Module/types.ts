export enum ModuleTypes {
  QUIZ = 'quiz',
  ARTICLE = 'article',
  PARENT = 'parent',
  EMULATOR = 'emulator',
  VALUE = 'value',
}

export interface ModuleInterface {
  id: number | string;
  moduleId: number | string;
  title: string;
  progress: number;
  type: ModuleTypes;
}