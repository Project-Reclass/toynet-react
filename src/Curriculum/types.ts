export interface Module {
  id: number;
  name: string;
  introduction: string;
  submodules: string;
}

export interface SubModule {
  id: number;
  name: string;
  introduction: string;
  type: 'SURVEY' | 'VALUE' | 'LESSON' | 'ARTICLE' | 'LAB';
}