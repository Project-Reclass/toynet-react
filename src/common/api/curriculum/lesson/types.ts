export interface LessonMetaRequest {
  id: number;
}