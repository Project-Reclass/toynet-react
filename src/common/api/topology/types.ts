/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
export type SessionId = string | number;

export interface TopologyResponse {
  id: number;
  author_id: number;
  topology: string;
}

export interface SessionRequest {
  toynet_topo_id: number;
  toynet_user_id: string;
}

export interface CommandRequest {
  id: SessionId;
  command: string;
}

export interface SessionRequestResponse {
  toynet_session_id: number;
}

export interface ToynetSession {
  topo_id: number;
  topology: string;
  user_id: string;
}

export interface ToynetSessionResponse {
  id: number;
  topology: string;
  user_id: number;
  image: string;
  time: string;
  last_update_time: string;
  toynetconfig: number;
}

export interface ToynetCommandResponse {
  output: string;
}

export interface ToyNetCreateHostRequest {
  ip: string;
  name: string;
  def_gateway: string;
}

export interface ToyNetCreateRouterRequest {
  ip: string;
  name: string;
  intfs: string[];
}

export interface ToyNetCreateSwitchRequest {
  name: string;
}

export interface ToyNetLinkRequest {
  dev_1: string;
  dev_2: string;
}

export interface ToyNetDeleteDeviceRequest {
  name: string;
}