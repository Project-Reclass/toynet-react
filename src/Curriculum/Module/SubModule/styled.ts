/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import styled from '@emotion/styled';
import { Box, Link, Text } from '@chakra-ui/react';

interface ModuleNameProps {
  locked: boolean;
  hoverColor?: string;
}

export const ModuleName = styled(Link)`
  color: ${({ locked }: ModuleNameProps) => locked ? 'grey' : 'white'};
  display: flex;
  cursor: pointer;
  justify-content: space-between;
  align-items: center;
  a {
    -webkit-user-select: none;
    user-select: none;
    svg {
      width: 1.5em;
      height: 1.5em;
    }
  }
  :hover {
    color: ${({ hoverColor }: ModuleNameProps) => hoverColor || 'white'};
  }
`;

interface SIBoxProps {
  type: 'lock' | 'check';
}

export const StatusIconBox = styled(Box)`
  background-color: ${({type}: SIBoxProps) => type !== 'check' ? 'white' : 'rgb(79, 149, 26)'};
  border-radius: 3px;
  width: 1.5rem;
  height: 1.5rem;
`;

export const SubmoduleDescription = styled(Text)`
  font-size: 14px;
  width: 80%;
  margin-left: 3rem;
`;