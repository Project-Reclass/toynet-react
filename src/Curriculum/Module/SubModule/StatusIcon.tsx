import React from 'react';
import { Stack, Icon } from '@chakra-ui/core';

import { StatusIconBox } from './styled';

interface SIProps {
  status: 'locked' | 'unlocked' | 'done'
}

const StatusIcon = ({ status }: SIProps) => (
  <Stack>
    <StatusIconBox type={status !== 'done' ? 'lock' : 'check'}>
      <Icon
        name={status === 'done' ? 'check' :
          status === 'locked' ? 'lock' : 'unlock'
        }
        margin='auto'
        marginTop='0.2rem'
        display='block'
        color={status !== 'done' ? 'grey' : 'white'}
      />
    </StatusIconBox>
  </Stack>
);


export default StatusIcon;
