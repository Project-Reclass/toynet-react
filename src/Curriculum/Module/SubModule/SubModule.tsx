/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import { FC } from 'react';
import { Divider, Flex, Link, Stack, Text, Collapse, Tooltip } from '@chakra-ui/react';
import { StarIcon } from '@chakra-ui/icons';
import { BsPlayCircle } from 'react-icons/bs';
import { SubModuleIntf } from 'src/common/types/curriculum';

import { ModuleName } from './styled';
import { useSessionStorage } from 'src/common/hooks/useSessionStorage';
import { createLink } from 'src/common/utils';
import { SubmoduleDescription } from './styled';

interface Props extends SubModuleIntf {
  moduleId: number;
  index: number;
  count: number;
}


const capitalize = (s: string): string =>
  `${s[0].toUpperCase()}${s.toLowerCase().slice(1)}`;

export const SubModule: FC<Props> = (
  {
    name,
    id,
    moduleId,
    type,
    index,
    count,
    introduction,
  },
) => {
  const [isOpen, setOpen] =
    useSessionStorage<boolean>(`submodule-${moduleId}-${id}-${index}`, true,
      value => JSON.parse(value));

  return (
    <Flex my='1'>
      <StarIcon
        w='1.5rem'
        h='1.5rem'
        color='yellow.400'
      />
      <Stack spacing={2} width='100%' marginLeft='1.5rem'>
        <Flex justifyContent='space-between'>
          <ModuleName locked={false}>
            <Tooltip
              hasArrow
              label={isOpen ? 'Show less' : 'Expand submodule'}
            >
              <Text onClick={() => setOpen(open => !open)}>
                {`${capitalize(type.toString())}: ${name}`}
              </Text>
            </Tooltip>
          </ModuleName>
          <ModuleName locked={false} hoverColor='rgba(84,143,155)'>
            <Link href={createLink({ moduleId, type, id })}>
              <BsPlayCircle data-testid='submodule_play_icon' />
            </Link>
          </ModuleName>
        </Flex>
        <Collapse in={isOpen} color='white'>
          <SubmoduleDescription>
            {introduction}
          </SubmoduleDescription>
        </Collapse>
        {index !== count -1 && <Divider />}
      </Stack>
    </Flex>
  );
};


export default SubModule;