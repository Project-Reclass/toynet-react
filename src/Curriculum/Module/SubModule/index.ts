import SubModule from './SubModule';
export default SubModule;