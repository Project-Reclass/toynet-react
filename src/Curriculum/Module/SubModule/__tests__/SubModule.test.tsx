/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/

/* eslint-disable no-magic-numbers */

import SubModule from 'src/Curriculum/Module/SubModule/SubModule';
import {  renderWithTheme } from 'src/common/test-utils/renderWithTheme';
import { SubModuleType } from 'src/common/types/curriculum';


const articleType = 'article' as SubModuleType;
const defaultProps = {
  name: 'article',
  title: 'Article',
  progress: 99,
  id: 1,
  moduleId: 2,
  type: articleType,
  introduction: 'this is an introduction',
  index: 0,
  count: 5,
};

describe('The sub modules', () => {
  it('should render and match previous snapshots', () => {
    const { container } = renderWithTheme(
      <SubModule {...defaultProps} />,
    );

    expect(container).toMatchSnapshot();
  });
  it('should create a link to the submodule page', () => {
    const { type, id, moduleId } = defaultProps;
    const { getByTestId } = renderWithTheme(<SubModule {...defaultProps} />);
    const playIcon = getByTestId('submodule_play_icon');

    expect(playIcon.closest('a')?.getAttribute('href')).toBe(`/module/${moduleId}/${type.toString()}/${id}`);
  });
});