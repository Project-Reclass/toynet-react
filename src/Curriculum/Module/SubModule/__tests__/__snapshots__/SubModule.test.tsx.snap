// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The sub modules should render and match previous snapshots 1`] = `
<div>
  <div
    class="css-1l0pjoz"
  >
    <svg
      class="chakra-icon css-1anit53"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M23.555,8.729a1.505,1.505,0,0,0-1.406-.98H16.062a.5.5,0,0,1-.472-.334L13.405,1.222a1.5,1.5,0,0,0-2.81,0l-.005.016L8.41,7.415a.5.5,0,0,1-.471.334H1.85A1.5,1.5,0,0,0,.887,10.4l5.184,4.3a.5.5,0,0,1,.155.543L4.048,21.774a1.5,1.5,0,0,0,2.31,1.684l5.346-3.92a.5.5,0,0,1,.591,0l5.344,3.919a1.5,1.5,0,0,0,2.312-1.683l-2.178-6.535a.5.5,0,0,1,.155-.543l5.194-4.306A1.5,1.5,0,0,0,23.555,8.729Z"
        fill="currentColor"
      />
    </svg>
    <div
      class="chakra-stack css-2y5ch"
    >
      <div
        class="css-gg4vpm"
      >
        <a
          class="chakra-link css-1e0lkyp"
        >
          <p
            class="chakra-text css-0"
          >
            Article: article
          </p>
        </a>
        <a
          class="chakra-link css-2izbe"
          hovercolor="rgba(84,143,155)"
        >
          <a
            class="chakra-link css-f4h6uy"
            href="/module/2/article/1"
          >
            <svg
              data-testid="submodule_play_icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
              />
              <path
                d="M6.271 5.055a.5.5 0 0 1 .52.038l3.5 2.5a.5.5 0 0 1 0 .814l-3.5 2.5A.5.5 0 0 1 6 10.5v-5a.5.5 0 0 1 .271-.445z"
              />
            </svg>
          </a>
        </a>
      </div>
      <div
        class="chakra-collapse"
        color="white"
        style="overflow: initial; display: block; opacity: 1; height: auto;"
      >
        <p
          class="chakra-text css-12txp0z"
        >
          this is an introduction
        </p>
      </div>
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-svjswr"
      />
    </div>
  </div>
</div>
`;
